root/static/scripts/edit/components/ExternalLinkAttributeDialog.js

```diff
@@ -0,0 +1,222 @@
+import type {LinkRelationshipT} from '../externalLinks';
+import DateRangeFieldset, {
+  partialDateFromField,
+  runReducer as runDateRangeFieldsetReducer,
+  type ActionT as DateRangeFieldsetActionT,
+} from './DateRangeFieldset';
+import {
+  createCompoundField,
+  createField,
+} from '../../edit/utility/createField';
+import {l} from '../../common/i18n';
+import {
+  applyAllPendingErrors,
+  hasSubfieldErrors,
+} from '../../../../utility/subfieldErrors';
+import {copyDatePeriodField} from '../utility/copyFieldData';
+
+type PropsT = {
+  onConfirm: (DatePeriodRoleT) => void,
+  relationship: LinkRelationshipT,
+};
+
+type StateT = {
+  +datePeriodField: DatePeriodFieldT,
+};
+
+type WritableStateT = {
+  ...StateT,
+  datePeriodField: WritableDatePeriodFieldT,
+};
+
+type ActionT =
+| {
+  +action: DateRangeFieldsetActionT,
+  +type: 'update-date-period',
+  }
+| {+type: 'reset'}
+| {+type: 'show-all-pending-errors'};
+
+const createInitialState = (props: PropsT): StateT => {
+  const relationship = props.relationship;
+  const beginDate = relationship.begin_date;
+  const endDate = relationship.end_date;
+
+  return {
+    datePeriodField: {
+      errors: [],
+      has_errors: false,
+      field: {
+        begin_date: createCompoundField(
+          'period.begin_date',
+          {
+            day: beginDate?.day ?? null,
+            month: beginDate?.month ?? null,
+            year: beginDate?.year ?? null,
+          },
+        ),
+        end_date: createCompoundField(
+          'period.end_date',
+          {
+            day: endDate?.day ?? null,
+            month: endDate?.month ?? null,
+            year: endDate?.year ?? null,
+          },
+        ),
+        ended: createField('period.ended', relationship.ended),
+      },
+      html_name: '',
+      id: 0,
+      type: 'compound_field',
+    },
+  };
+};
+
+const ExternalLinkAttributeDialog = (props: PropsT): React.MixedElement => {
+  const buttonRef = React.useRef<HTMLButtonElement | null>(null);
+  const [open, setOpen] = React.useState(false);
+  const initialState = createInitialState(props);
+
+  const resetReducer = (state: WritableStateT): void => {
+    copyDatePeriodField(initialState.datePeriodField, state.datePeriodField);
+  };
+
+  const reducer = (state: StateT, action: ActionT): StateT => {
+    return mutate<WritableStateT, StateT>(
+      state, (newState) => {
+        switch (action.type) {
+          case 'update-date-period':
+            runDateRangeFieldsetReducer(
+              newState.datePeriodField,
+              action.action,
+            );
+            break;
+          case 'reset':
+            resetReducer(newState);
+            break;
+          case 'show-all-pending-errors':
+            applyAllPendingErrors(newState.datePeriodField);
+            break;
+          default:
+            throw new Error('Unknown action: ' + action.type);
+        }
+      },
+    );
+  };
```

Comment:
I would store props.relationship in state and move these functions outside the component so each render can reuse them.